### PR TITLE
fix: got hooks working

### PR DESCRIPTION
**Pull request by markusahlstrand**
Add a hook for linking users. Only removed the old explicit linking in one place so far to minimize the impact but will go through and remove the other places as well.

### CODE REVIEW

File: src/authentication-flows/social.tsx [hunk 1]
```diff
@@ -208,13 +208,7 @@ export async function socialAuthCallback({
...
     ctx.set("logType", LogTypes.SUCCESS_SIGNUP);
 
-    const primaryUser = await getPrimaryUserByEmail({
-      userAdapter: env.data.users,
-      tenant_id: client.tenant_id,
-      email: email,
-    });
-
-    const newSocialUser = await env.data.users.create(client.tenant_id, {
+    user = await env.data.users.create(client.tenant_id, {
```
Comment (danielgent) [inline src/authentication-flows/social.tsx:211]: I never liked this mutable reassignment (it already existed though so it's correct to keep it here :+1: )

File: src/hooks/link-users.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { User } from "../types";
+import { DataAdapters } from "../adapters/interfaces";
+import { getPrimaryUserByEmail } from "../utils/users";
+
+export function linkUsersHook(data: DataAdapters) {
+  return async (tenant_id: string, user: User): Promise<User> => {
+    // If the user does not have an email or the email is not verified, return the user
+    if (!user.email || !user.email_verified) {
```
Comment (danielgent) [inline src/hooks/link-users.ts:8]: :+1: Nice. This is how you solve this. Do the `user.email_verified` check in the create hook :+1:

File: test/integration/flows/social.spec.ts [hunk 3]
```diff
@@ -554,84 +554,40 @@ describe("social sign on", () => {
...
     });
+
     it("should return existing primary account when logging in with new social sign on with same email address AND there is already another linked social account", async () => {
-      const token = await getAdminToken();
       const env = await getEnv();
       const oauthClient = testClient(oauthApp, env);
-      const managementClient = testClient(managementApp, env);
 
-      // What I want here is for the linked social account to be returned FIRST!
-      // so this is a bit synthetic by manually writing the users...
```
Comment (danielgent) [inline test/integration/flows/social.spec.ts:564]: Why did I do the test like this? I'm not testing the actual linking of users :smile: 

Apologies for this! I imagine it was a nightmare to work out :see_no_evil: 

This is much more correct now. You create the primary account first, then the social account next links to this


We must have had some weird bug for me to write something so complex...  It gives me confidence in the tests though if I wrote such edge case tests

File: src/hooks/link-users.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    if (!primaryUser) {
+      return data.users.create(tenant_id, user);
+    }
+
+    await data.users.create(tenant_id, {
+      ...user,
+      linked_to: primaryUser.id,
+    });
+
+    // TODO: add the new user to the identities
```
Comment (danielgent) [inline src/hooks/link-users.ts:29]: I'm surprised we don't have a test that is asserting this :see_no_evil:  Maybe we weren't even doing this before

**Review by danielgent — APPROVED** [review]
:checkered_flag:
